rpc/src/v1/impls/secretstore.rs

```diff
@@ -0,0 +1,128 @@
+pub struct SecretStoreClient {
+	accounts: Option<Weak<AccountProvider>>,
+}
+
+impl SecretStoreClient {
+	/// Creates new SecretStoreClient
+	pub fn new(store: &Option<Arc<AccountProvider>>) -> Self {
+		SecretStoreClient {
+			accounts: store.as_ref().map(Arc::downgrade),
+		}
+	}
+
+	/// Attempt to get the `Arc<AccountProvider>`, errors if provider was not
+	/// set, or if upgrading the weak reference failed.
+	fn account_provider(&self) -> Result<Arc<AccountProvider>, Error> {
+		unwrap_provider(&self.accounts)
+	}
+
+	/// Decrypt public key using account' private key
+	fn decrypt_key(&self, address: H160, key: Bytes) -> Result<Vec<u8>, Error> {
+		let store = self.account_provider()?;
+		store.decrypt(address.into(), None, &DEFAULT_MAC, &key.0)
+			.map_err(|e| errors::account("Could not decrypt key.", e))
+	}
+
+	/// Decrypt secret key using account' private key
+	fn decrypt_secret(&self, address: H160, key: Bytes) -> Result<Secret, Error> {
+		self.decrypt_key(address, key)
+			.and_then(|s| Secret::from_slice(&s).map_err(|e| errors::account("invalid secret", e)))
+	}
+}
+
+impl SecretStore for SecretStoreClient {
+	fn encrypt(&self, address: H160, key: Bytes, data: Bytes) -> Result<Bytes, Error> {
+		encryption::encrypt_document(self.decrypt_key(address, key)?, data.0)
+			.map(Into::into)
+	}
+
+	fn decrypt(&self, address: H160, key: Bytes, data: Bytes) -> Result<Bytes, Error> {
+		encryption::decrypt_document(self.decrypt_key(address, key)?, data.0)
+			.map(Into::into)
+	}
+
+	fn shadow_decrypt(&self, address: H160, decrypted_secret: H512, common_point: H512, decrypt_shadows: Vec<Bytes>, data: Bytes) -> Result<Bytes, Error> {
+		let mut shadows = Vec::with_capacity(decrypt_shadows.len());
+		for decrypt_shadow in decrypt_shadows {
+			shadows.push(self.decrypt_secret(address.clone(), decrypt_shadow)?);
+		}
+
+		encryption::decrypt_document_with_shadow(decrypted_secret.into(), common_point.into(), shadows, data.0)
+			.map(Into::into)
+	}
+}
+
+#[cfg(not(feature="secretstore"))]
+mod encryption {
+	use ethkey::{Secret, Public};
+	use jsonrpc_core::Error;
+	use util::Bytes;
+	use v1::helpers::errors;
+
+	pub fn encrypt_document(_key: Vec<u8>, _document: Bytes) -> Result<Bytes, Error> {
+		Err(errors::secretstore_disabled())
+	}
+
+	pub fn decrypt_document(_key: Vec<u8>, _document: Bytes) -> Result<Bytes, Error> {
+		Err(errors::secretstore_disabled())
+	}
+
+	pub fn decrypt_document_with_shadow(_decrypted_secret: Public, _common_point: Public, _shadows: Vec<Secret>, _document: Bytes) -> Result<Bytes, Error> {
+		Err(errors::secretstore_disabled())
+	}
+}
+
+#[cfg(feature="secretstore")]
+mod encryption {
+	use ethkey::{Secret, Public};
+	use jsonrpc_core::Error;
+	use ethcore_secretstore;
+	use util::Bytes;
+	use v1::helpers::errors;
+
+	pub fn encrypt_document(key: Vec<u8>, document: Bytes) -> Result<Bytes, Error> {
+		ethcore_secretstore::encrypt_document(key, document)
+			.map_err(|e| errors::encryption_error(e))
```

Comment:
`map_err(errors::encryption_error)` should be enough. (You should also ditch the `_error` suffix)